File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/NativeUI/SafeArea/SafeAreaNMM.cs
namespace DwebBrowser.MicroService.Browser.NativeUI.SafeArea;

public class SafeAreaNMM : NativeMicroModule
{
    public SafeAreaNMM() : base("safe-area.nativeui.browser.dweb")
    {
    }

    private Task<SafeAreaController> _getControllerAsync(Mmid mmid) =>
        MainThread.InvokeOnMainThreadAsync(() => NativeUiController.FromMultiWebView(mmid).SafeArea);

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        HttpRouter.AddRoute(IpcMethod.Get, "/getState", async (_, ipc) =>
        {
            return await _getControllerAsync(ipc.Remote.Mmid);
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/setState", async (request, ipc) =>
        {
            var controller = await _getControllerAsync(ipc.Remote.Mmid);
            request.QueryBool("overlay")?.Also(it =>
            {
                controller.OverlayState.Set(it);
                controller.Observer.Get();
            });
            return null;
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/startObserve", async (_, ipc) =>
        {
            return (await _getControllerAsync(ipc.Remote.Mmid)).StateObserver.StartObserve(ipc);
        });

        HttpRouter.AddRoute(IpcMethod.Get, "/stopObserve", async (_, ipc) =>
        {
            return (await _getControllerAsync(ipc.Remote.Mmid)).StateObserver.StopObserve(ipc);
        });
    }
}

File: android/app/src/main/java/info/bagen/dwebbrowser/microService/browser/nativeui/safeArea/SafeAreaNMM.kt
package info.bagen.dwebbrowser.microService.browser.nativeui.safeArea

import info.bagen.dwebbrowser.microService.browser.nativeui.NativeUiController
import info.bagen.dwebbrowser.microService.browser.nativeui.helper.QueryHelper
import info.bagen.dwebbrowser.microService.browser.nativeui.helper.fromMultiWebView
import org.dweb_browser.microservice.core.BootstrapContext
import org.dweb_browser.microservice.core.NativeMicroModule
import org.dweb_browser.microservice.help.cors
import org.dweb_browser.microservice.help.types.MICRO_MODULE_CATEGORY
import org.dweb_browser.microservice.help.types.MMID
import org.http4k.core.Method
import org.http4k.core.Response
import org.http4k.core.Status
import org.http4k.routing.bind
import org.http4k.routing.routes

class SafeAreaNMM : NativeMicroModule("safe-area.nativeui.browser.dweb", "safeArea") {

  init {
    categories = listOf(MICRO_MODULE_CATEGORY.Service, MICRO_MODULE_CATEGORY.Render_Service);
  }

  private fun getController(mmid: MMID) = NativeUiController.fromMultiWebView(mmid).safeArea

  override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
    apiRouting = routes(
      /** 获取状态 */
      "/getState" bind Method.GET to defineHandler { _, ipc ->
        return@defineHandler getController(ipc.remote.mmid)
      },
      /** 获取状态 */
      "/setState" bind Method.GET to defineHandler { request, ipc ->
        val controller = getController(ipc.remote.mmid)
        QueryHelper.overlay(request)?.also { controller.overlayState.value = it }
        return@defineHandler null
      },
      /**
       * 开始数据订阅
       */
      "/observe" bind Method.GET to defineHandler { _, ipc ->
        val inputStream = getController(ipc.remote.mmid).observer.startObserve(ipc)
        return@defineHandler Response(Status.OK).body(inputStream)
      },
    ).cors()
  }

  override suspend fun _shutdown() {
  }
}
